crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,410 @@
+            next
+          end
+          nodes_map[node_name] = [] if nodes_map[node_name].nil?
+          nodes_map[node_name] << role_name
+        end
+      end
+
+      nodes_map
+    end
+
+    # Get a hash of {node => [roles], node1 => [roles]}
+    def add_pending_elements(bc, inst, elements, queue_me, pre_cached_nodes = {})
+      nodes_map = elements_to_nodes_to_roles_map(elements)
+
+      # We need to be sure that we're the only ones modifying the node records at this point.
+      # This will work for preventing changes from rails app, but not necessarily chef.
+      # Tough luck.
+      f = file_lock.acquire("BA-LOCK", logger: logger)
+
+      # Delay is the list of nodes that are not ready and are needed for this deploy to run
+      delay = []
+      pre_cached_nodes = {}
+      begin
+        # Check for delays and build up cache
+        # FIXME: why?
+        if queue_me
+          delay = nodes_map.keys
+        else
+          delay, pre_cached_nodes = elements_not_ready(nodes_map.keys, pre_cached_nodes)
+        end
+
+        # We have all nodes we'll ever need. Mark them as applying and this proposal
+        # as the 'author'
+        if delay.empty?
+          nodes_map.each do |node_name, val|
+            node = pre_cached_nodes[node_name]
+
+            # Nothing to wait for so mark them applying now.
+            node.crowbar['state'] = 'applying'
+            node.crowbar['state_owner'] = "#{bc}-#{inst}"
+            node.save
+          end
+        else
+          # Update all nodes affected by this proposal deploy (elements) -> add info that this proposal
+          # will add list of roles to node's crowbar.pending hash.
+          nodes_map.each do |node_name, val|
+            # Make sure we have a node.
+            node = pre_cached_nodes[node_name]
+            node = NodeObject.find_node_by_name(node_name) if node.nil?
+            next if node.nil?
+            pre_cached_nodes[node_name] = node
+
+            # Mark node as pending. User will be informed about node needing
+            # manual allocation if not allocated.
+            node.crowbar["crowbar"]["pending"] = {} if node.crowbar["crowbar"]["pending"].nil?
+            node.crowbar["crowbar"]["pending"]["#{bc}-#{inst}"] = val
+            node.save
+          end
+        end
+      rescue StandardError => e
+        logger.fatal("add_pending_elements: Exception #{e.message} #{e.backtrace.join("\n")}")
+      ensure
+        file_lock.release(f)
+      end
+
+      [ delay, pre_cached_nodes ]
+    end
+
+    # Assumes the BA-LOCK is held
+    def elements_not_ready(nodes, pre_cached_nodes = {})
+      # Check to see if we should delay our commit until nodes are ready.
+      delay = []
+      nodes.each do |n|
+        node = NodeObject.find_node_by_name(n)
+        next if node.nil?
+
+        pre_cached_nodes[n] = node
+        delay << n if node.crowbar['state'] != "ready" and !delay.include?(n)
```

Comment:
Style/AndOr: Use `&&` instead of `and`.<br>Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.